### ISSUE TITLE
'must be size_t' error with gcc on Windows

### ISSUE THREAD

**Issue by bigbigrain** [microsoft/vscode-cpptools#6573]
Type: Debugger

**Describe the bug**
- OS and Version: Windows 10
- VS Code Version: 1.51.1
- C/C++ Extension Version: 1.1.2
- Other extensions you installed (and if the issue persists after disabling them): i disabled all other extensions.
- A clear and concise description of what the bug is.
When i use "https://github.com/nlohmann/json" this c++ lib, using their user defined literal "_json" it will throw a problem,say:user-defined literal operator not found.

**To Reproduce**
*Please include a code sample and `launch.json` configuration.*
```
"launch": {
		"version": "0.2.0",
		"configurations": [

			{
				"name": "Cpp",
				"type": "cppdbg",
				"request": "launch",
				"program": "${fileDirname}/${fileBasenameNoExtension}.exe",
				"args": [],
				"stopAtEntry": false,
				"cwd": "${fileDirname}",
				"environment": [],
				"externalConsole": false,
				"internalConsoleOptions": "neverOpen",
				"MIMode": "gdb",
				"miDebuggerPath": "D:\\environment\\gcc\\bin\\gdb.exe",
				"setupCommands": [{
					"description": "Enable pretty-printing for gdb",
					"text": "-enable-pretty-printing",
					"ignoreFailures": false
				}],
				"preLaunchTask": "cpp-gcc"
			},
			{
				"name": "C",
				"type": "cppdbg",
				"request": "launch",
				"program": "${fileDirname}/${fileBasenameNoExtension}.exe",
				"args": [],
				"stopAtEntry": false,
				"cwd": "${fileDirname}",
				"environment": [],
				"externalConsole": false,
				"internalConsoleOptions": "neverOpen",
				"MIMode": "gdb",
				"miDebuggerPath": "gdb.exe",
				"setupCommands": [{
					"description": "Enable pretty-printing for gdb",
					"text": "-enable-pretty-printing",
					"ignoreFailures": false
				}],
				"preLaunchTask": "c-gcc"
			}
		]
	}

"tasks": {
		"version": "2.0.0",
		"tasks": [{
				"label": "cpp-gcc",
				"command": "g++",
				"args": [
					"${file}",
					"-o",
					"${fileDirname}/${fileBasenameNoExtension}.exe",
					"-g",
					"-Wall",
					"-static-libgcc",
					"-std=c++2a",
					"-I", "D:/Code/C++/head"
				],
				"type": "process",
				"group": {
					"kind": "build",
					"isDefault": true
				},
				"presentation": {
					"echo": true,
					"reveal": "always",
					"focus": false,
					"panel": "shared"
				}
			},
			{
				"label": "c-gcc",
				"command": "gcc",
				"args": [
					"${file}",
					"-o",
					"${fileDirname}/${fileBasenameNoExtension}.exe",
					"-g",
					"-Wall",
					"-static-libgcc",
					"-std=c17"
				],
				"type": "process",
				"group": {
					"kind": "build",
					"isDefault": true
				},
				"presentation": {
					"echo": true,
					"reveal": "always",
					"focus": false,
					"panel": "shared"
				},
				"problemMatcher": [
					"$gcc"
				]
			}
		]
	},
```

example code:
```
 #include <D:\\Code\\C++\\head\\json\\json.hpp>
 //json.hpp is https://github.com/nlohmann/json       //v3.9.1
 #include <iostream>

 using json = nlohmann::json;
 
 int main() {
    json js = "{ \"happy\": [true,false,true,true], \"pi\": 3.14159 }"_json;
     std::cout << js["pi"] << std::endl;
    return 0;
 }
```
1. write the code
2. Static syntax hint will warning me line " json js = "{ \"happy\": [true,false,true,true], \"pi\": 3.14159 }"_json;" has user-defined literal operator not found
3. program is Compile and run NORMALLY, but vscode keep showing error.
(i'm a tiro in learn c++ lang, if something wrong. I am very sorry. )
<img width="849" alt="SC-20-11-24-213938" src="https://user-images.githubusercontent.com/17808375/100103794-545da380-2ea0-11eb-80a0-08aca1f2c0da.png">
<img width="634" alt="SC-20-11-24-213901" src="https://user-images.githubusercontent.com/17808375/100103809-59225780-2ea0-11eb-9ba2-91f5ffd9410c.png">

**Additional context**
*If applicable, please include logging by adding "logging": { "engineLogging": true, "trace": true, "traceResponse": true } in your `launch.json`*
Add any other context about the problem here including log or error messages in your Debug Console or Output windows.
.

**Comment by Colengms** [comment]
Hi @bigbigrain .  Since you are using gcc on Windows, are you using MinGW or another gcc compiler?  There are some issues with gcc on Windows due to different implementations handling int sizes differently.  We are currently tracking this with https://github.com/microsoft/vscode-cpptools/issues/5457 (as it was previously reported due to use of size_t in operator new).

If you are using MinGW, I would suggest using the latest version (10.2.0), which is available as a package for MSYS2.  Once installed, it can be referenced directly in VS Code.  It doesn't require being run within MSYS2.  I believe that build of gcc would match the int sizes we are expecting.

**Comment by mcilloni** [comment]
I'm also having the same issue with GCC 8 and Xtensa (ESP32):

![Screenshot_20210217_105622](https://user-images.githubusercontent.com/424549/108187455-c85dbf80-710e-11eb-816d-f7605fef862d.png)

This is obviously wrong, given that size_t IS unsigned int, and IntelliSense known that:

![Screenshot_20210217_105715](https://user-images.githubusercontent.com/424549/108187588-e9261500-710e-11eb-9d33-3efa78a08a74.png)

Everything else works fine.
The compiler is GCC 8.4, triple is xtensa-esp32-elf, `sizeof(int) == 4`.

**Comment by sean-mcmanus** [comment]
@mcilloni What OS and intelliSenseMode are you using?

**Comment by mcilloni** [comment]
The OS is Linux, and I'm not aware of what intelliSenseMode is being set because I'm using the CMake extension as the configuration provider.

If I set logging to to info, the C/C++ extension prints this:

```
Compiler query command line: "/home/marco/.local/bin/xtensa-esp32-elf-gcc" -mlongcalls -Wno-frame-address -fdiagnostics-color=always -ffunction-sections -fdata-sections -fstrict-volatile-bitfields -nostdlib -Wall -Werror=all -Wno-error=unused-function -Wno-error=unused-but-set-variable -Wno-error=unused-variable -Wno-error=deprecated-declarations -Wextra -Wno-unused-parameter -Wno-sign-compare -ggdb -Og -std=gnu++11 -fexceptions -fno-rtti -std=c++17  -Wp,-v -E -dD -x c++ -m64  /dev/null
Failed to query compiler. Falling back to 32-bit intelliSenseMode.
Compiler query command line: "/home/marco/.local/bin/xtensa-esp32-elf-gcc" -mlongcalls -Wno-frame-address -fdiagnostics-color=always -ffunction-sections -fdata-sections -fstrict-volatile-bitfields -nostdlib -Wall -Werror=all -Wno-error=unused-function -Wno-error=unused-but-set-variable -Wno-error=unused-variable -Wno-error=deprecated-declarations -Wextra -Wno-unused-parameter -Wno-sign-compare -ggdb -Og -std=gnu++11 -fexceptions -fno-rtti -std=c++17  -Wp,-v -E -dD -x c++ -m32  /dev/null
Failed to query compiler. Falling back to no bitness.
```

**Comment by sean-mcmanus** [comment]
You can find the intelliSenseMode via running the C/C++: Log Diagnostics command. Can you provde that output? This might be an issue with the configuration info we're receiving from CMake Tools.

**Comment by mcilloni** [comment]
> You can find the intelliSenseMode via running the C/C++: Log Diagnostics command. Can you provde that output? This might be an issue with the configuration info we're receiving from CMake Tools.

Yep, I run that command and it's clear that the wrong mode is being picked (gcc-x64 instead of gcc-x86, which if I remember correctly is the right mode for 32 bit):

```
-------- Diagnostics - 2/22/2021, 2:25:17 PM
Version: 1.2.1
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [],
    "compilerPath": "/home/marco/.local/bin/clang",
    "cStandard": "c17",
    "cppStandard": "c++14",
    "intelliSenseMode": "linux-clang-x64",
    "compilerArgs": [],
    "intelliSenseModeIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "compilerPathIsExplicit": true,
    "configurationProvider": "ms-vscode.cmake-tools",
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
Custom browse configuration: 
{
    "browsePath": [
          // ... list of a bajilion include directories ...
    ],
    "standard": "gnu++11",
    "compilerPath": "/home/marco/.local/bin/xtensa-esp32-elf-gcc",
    "compilerArgs": [
        "-mlongcalls",
        "-Wno-frame-address",
        "-fdiagnostics-color=always",
        "-ffunction-sections",
        "-fdata-sections",
        "-fstrict-volatile-bitfields",
        "-nostdlib",
        "-Wall",
        "-Werror=all",
        "-Wno-error=unused-function",
        "-Wno-error=unused-but-set-variable",
        "-Wno-error=unused-variable",
        "-Wno-error=deprecated-declarations",
        "-Wextra",
        "-Wno-unused-parameter",
        "-Wno-sign-compare",
        "-ggdb",
        "-Og",
        "-std=gnu++11",
        "-fexceptions",
        "-fno-rtti",
        "-D_GNU_SOURCE",
        "-DIDF_VER=\"a4b1e19cb\"",
        "-DGCC_NOT_5_2_0",
        "-DESP_PLATFORM",
        "-std=gnu++11"
    ]
}
Custom configurations:
[ /home/marco/Workspace/p501_esp32/components/p501_fsm/fsm_event_handler.cpp ]
{
    "defines": [
        "MBEDTLS_CONFIG_FILE=\"mbedtls/esp_config.h\"",
        "_GNU_SOURCE",
        "IDF_VER=\\\"a4b1e19cb-dirty\\\"",
        "GCC_NOT_5_2_0",
        "ESP_PLATFORM"
    ],
    "standard": "c++17",
    "includePath": [
        // omitted
    ],
    "intelliSenseMode": "gcc-x64", <- WRONG, not 64 bit
    "compilerPath": "/home/marco/.local/bin/xtensa-esp32-elf-gcc",
    "compilerArgs": [
        "-mlongcalls",
        "-Wno-frame-address",
        "-fdiagnostics-color=always",
        "-ffunction-sections",
        "-fdata-sections",
        "-fstrict-volatile-bitfields",
        "-nostdlib",
        "-Wall",
        "-Werror=all",
        "-Wno-error=unused-function",
        "-Wno-error=unused-but-set-variable",
        "-Wno-error=unused-variable",
        "-Wno-error=deprecated-declarations",
        "-Wextra",
        "-Wno-unused-parameter",
        "-Wno-sign-compare",
        "-ggdb",
        "-Og",
        "-std=gnu++11",
        "-fexceptions",
        "-fno-rtti",
        "-D_GNU_SOURCE",
        "-DIDF_VER=\\\"a4b1e19cb-dirty\\\"",
        "-DGCC_NOT_5_2_0",
        "-DESP_PLATFORM",
        "-std=c++17"
    ]
}
```

And then, below,

```
Translation Unit Configurations:
[ /home/marco/Workspace/p501_esp32/components/p501_fsm/fsm_event_handler.cpp ]:
    Process ID: 100699
    Memory Usage: 280 MB
    Compiler Path: /home/marco/.local/bin/xtensa-esp32-elf-gcc
    Includes:
        // cut
    Defines:
        MBEDTLS_CONFIG_FILE="mbedtls/esp_config.h"
        _GNU_SOURCE
        IDF_VER=\"a4b1e19cb-dirty\"
        GCC_NOT_5_2_0
        ESP_PLATFORM
    Standard Version: c++17
    IntelliSense Mode: linux-gcc-x64 <- WRONG!
    Other Flags:
        --g++
        --gnu_version=80400
Total Memory Usage: 509 MB

// cut

```

**Comment by sean-mcmanus** [comment]
@Colengms @elahehrashedi Is this a CMake Tools bug?

**Comment by Colengms** [comment]
Since @mcilloni has a different scenario, I'll move it to a new issue.

**Comment by sean-mcmanus** [comment]
I believe the original issue (Windows + mingw clang/gcc) should be fixed with 1.3, same root cause as https://github.com/microsoft/vscode-cpptools/issues/7016

**Comment by sean-mcmanus** [comment]
This should be fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/1.3.0-insiders . Let us know if you're still hitting an issue.

**Comment by michelleangela** [comment]
Fix is released in the 1.3.0 release version https://github.com/microsoft/vscode-cpptools/releases/tag/1.3.0
